Cesium3DTilesSelection/src/Tileset.cpp

```diff
@@ -294,6 +310,127 @@ Tileset::updateViewOffline(const std::vector<ViewState>& frustums) {
+
+    // If any candidates need loading, add to return set
+    for (Tile* pTile : query.candidateTiles) {
+      if (pTile->getState() != TileLoadState::Done) {
+        tilesNeedingLoading.insert(pTile);
+        tileStillNeedsLoading = true;
+      }
+    }
+  }
+
+  // Bail if we're waiting on tiles to load
+  if (tileStillNeedsLoading)
+    return false;
+
+  // Do the intersect tests
+  for (TerrainQuery& query : request.queries) {
+    for (Tile* pTile : query.candidateTiles)
+      query.intersectVisibleTile(pTile);
+  }
+
+  // All rays are done, create results
+  Tileset::HeightResults results;
+
+  // Start with any warnings from tile traversal
+  results.warnings = std::move(warnings);
+
+  // Populate results with completed queries
+  for (TerrainQuery& query : request.queries) {
+    Tileset::HeightResults::CoordinateResult coordinateResult = {
+        query.intersectResult.hit.has_value(),
+        std::move(query.inputCoordinate)};
+
+    // Add query warnings into the height result
+    std::copy(
+        query.intersectResult.warnings.begin(),
+        query.intersectResult.warnings.end(),
+        std::back_inserter(results.warnings));
+
+    if (coordinateResult.heightAvailable)
+      coordinateResult.coordinate.height =
+          RAY_ORIGIN_HEIGHT -
+          glm::sqrt(query.intersectResult.hit->rayToWorldPointDistanceSq);
+
+    results.coordinateResults.push_back(coordinateResult);
+  }
+
+  request.promise.resolve(std::move(results));
+  return true;
+}
+
+void Tileset::visitHeightRequests() {
+  if (_heightRequests.size() == 0)
+    return;
+
+  // Go through all requests, either complete them, or gather the tiles they
+  // need for completion
+  std::set<Tile*> tilesNeedingLoading;
+  for (auto it = _heightRequests.begin(); it != _heightRequests.end();) {
+    HeightRequest& request = *it;
+    if (!tryCompleteHeightRequest(request, tilesNeedingLoading)) {
+      ++it;
+    } else {
+      auto deleteIt = it;
+      ++it;
+      _heightRequests.erase(deleteIt);
+    }
+  }
+
+  // Add a load request for tiles that are needed, and haven't started
+  TileLoadPriorityGroup priorityGroup = TileLoadPriorityGroup::Urgent;
+  double priority = 0.0;
+  for (Tile* pTile : tilesNeedingLoading) {
+    TileLoadState loadState = pTile->getState();
+
+    CESIUM_ASSERT(loadState != TileLoadState::Done);
+
+    // Push to appropriate queue based on state
+    if (loadState == TileLoadState::Unloaded) {
+
+      // Only add if not already in queue
+      auto foundIt = std::find_if(
+          this->_workerThreadLoadQueue.begin(),
+          this->_workerThreadLoadQueue.end(),
+          [pTile](const TileLoadTask& task) { return task.pTile == pTile; });
+
+      if (foundIt == this->_workerThreadLoadQueue.end())
+        this->_workerThreadLoadQueue.push_back(
+            {pTile, priorityGroup, priority});
+
+    } else if (loadState == TileLoadState::ContentLoaded) {
+
+      if (pTile->isRenderContent()) {
```

Comment:
Performance question, do we absolutely need to transition this from `ContentLoaded` to `ContentDone`?

Can we perform the intersection tests in this state, to save waiting for all the things that are done in `finishLoading`.